### PR TITLE
#6946 Run hooks on merge/edit and cope with protected branches

### LINKED ISSUE #6003: no git hook executed during "gitea squash and merge"
- Gitea version (or commit ref): dfad569, 1.7.1 (docker image)
- Git version: git version 2.18.1
- Operating system: Alpine Linux 3.8.2
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist:

## Description

running a "squash and merge" in a pull request after an approval does not trigger a git hook post-receive. The git hook script is a bash script provided with "set -x". No corresponding output is found in any logfile.
But the merge is done.
Doing a git merge without gitea+pull request works like a charm. 

## Screenshots
none.

### LINKED ISSUE #6946: Error when merge pull request
- Gitea version (or commit ref): 1.9.0+dev-249-g8b36f01f4
- Git version: 2.20.1
- Operating system: docker
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist:

```txt

2019/05/14 12:11:57 ...orm/session_exist.go:79:Exist() [I] [SQL] SELECT "id", "org_id", "team_id", "uid" FROM "team_user" WHERE (uid=$1) AND "team_id" IN ($2) LIMIT 1 []interface {}{2, 28},
[Macaron] 2019-05-14 12:11:57: Started GET /api/internal/branch/69/master for 127.0.0.1,
2019/05/14 12:11:57 modules/auth/auth.go:97:CheckOAuthAccessToken() [T] ParseOAuth2Token: signature is invalid,
2019/05/14 12:11:57 ...s/context/context.go:308:func1() [D] Session ID: dad04db9e123606d,
2019/05/14 12:11:57 ...s/context/context.go:309:func1() [D] CSRF Token: o5QUfe4WWRuQbs-M3v_LlmRjt2I6MTU1NzgzNTkxNzIxMTM1MDExMA==,
2019/05/14 12:11:57 .../xorm/session_get.go:95:nocacheGet() [I] [SQL] SELECT "id", "repo_id", "branch_name", "can_push", "enable_whitelist", "whitelist_user_i_ds", "whitelist_team_i_ds", "enable_merge_whitelist", "merge_whitelist_user_i_ds", "merge_whitelist_team_i_ds", "approvals_whitelist_user_i_ds", "approvals_whitelist_team_i_ds", "required_approvals", "created_unix", "updated_unix" FROM "protected_branch" WHERE "repo_id"=$1 AND "branch_name"=$2 LIMIT 1 []interface {}{69, "master"},
2019/05/14 12:11:57 routers/repo/pull.go:605:MergePullRequest() [E] Merge: git push: From /data/git/repositories/project/shell-asset,
[Macaron] 2019-05-14 12:11:57: Completed GET /api/internal/branch/69/master 200 OK in 2.41096ms,
	 * [new branch]      feat_archive -> head_repo/feat_archive,
	 * [new branch]      fixBug       -> head_repo/fixBug,
	 * [new branch]      master       -> head_repo/master,
	Automatic merge went well; stopped before committing as requested,
	remote: Gitea: Internal error        ,
	remote: Fail to detect force push: chdir /data/git/repositories/shell-asset.git: no such file or directory        ,
	To /data/git/repositories/project/shell-asset.git,
	 ! [remote rejected] master -> master (pre-receive hook declined),
	error: failed to push some refs to '/data/git/repositories/project/shell-asset.git',
2019/05/14 12:11:57 models/webhook.go:867:DeliverHooks() [T] DeliverHooks [repo_id: 69],
2019/05/14 12:11:57 ...s/context/context.go:138:HTML() [D] Template: status/500,
2019/05/14 12:11:57 models/pull.go:1298:AddTestPullRequestTask() [T] AddTestPullRequestTask [head_repo_id: 69, head_branch: master]: finding pull requests,
[Macaron] 2019-05-14 12:11:57: Completed POST /Project/Shell-Asset/pulls/83/merge 500 Internal Server Error in 158.425011ms
```

## Description

When I merge a pull request of repository `Project/Shell-Asset`, I got a 500 error. `Project` is a private organization and `shell-asset` is private repository also. I disabled git hook from app.ini but does not work.

### PR DESCRIPTION
**Pull request by zeripath**
Fix #6946 by setting and checking PullRequest ID on pushing if we're pushing from a PR.
Fixes #6003.

### CODE REVIEW

**Comment by zeripath** [comment]
OK two things:

1. I'm getting a deadlock on merges which is interesting because I don't think that should be happening.
2. I'm also getting an error on the tests.

Right I think I'll leave this as WIP but open a second PR to stop running the push hooks on merge.

**Comment by zeripath** [comment]
OK I've discovered the issue here, TestSearchRepos is broken.

**Comment by zeripath** [comment]
Now working after applying #7004 to fix that broken test

File: integrations/repo_search_test.go [hunk 1]
```diff
@@ -27,6 +31,23 @@ func resultFilenames(t testing.TB, doc *HTMLDoc) []string {
...
+	repo, err := models.GetRepositoryByOwnerAndName("user2", "repo1")
+	assert.NoError(t, err)
+
+	models.UpdateRepoIndexer(repo)
+
+	log.Printf("Waiting for indexing\n")
+
+	i := 0
+	for {
+		if repo.IndexerStatus != nil && len(repo.IndexerStatus.CommitSha) != 0 && i <= 60 {
```
Comment (zeripath) [inline integrations/repo_search_test.go]:
```suggestion
		if (repo.IndexerStatus != nil && len(repo.IndexerStatus.CommitSha) != 0) || i > 60 {
```

File: modules/private/hook.go [hunk 2]
```diff
@@ -29,11 +29,12 @@ type HookOptions struct {
 	UserName                        string
 	GitObjectDirectory              string
 	GitAlternativeObjectDirectories string
+	ProtectedBranchID               int64
 }
 
 // HookPreReceive check whether the provided commits are allowed
 func HookPreReceive(ownerName, repoName string, opts HookOptions) (int, string) {
-	reqURL := setting.LocalURL + fmt.Sprintf("api/internal/hook/pre-receive/%s/%s?old=%s&new=%s&ref=%s&userID=%d&gitObjectDirectory=%s&gitAlternativeObjectDirectories=%s",
+	reqURL := setting.LocalURL + fmt.Sprintf("api/internal/hook/pre-receive/%s/%s?old=%s&new=%s&ref=%s&userID=%d&gitObjectDirectory=%s&gitAlternativeObjectDirectories=%s&prID=%d",
```
Comment (lunny) [inline modules/private/hook.go:37]: Could we get the `prID` from routers but not envs?

Comment (zeripath) [reply modules/private/hook.go:37]: Unfortunately not. Remember the hooks are called from within git.

**Comment by lunny** [comment]
Please resolve conflicts

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/pull.go [hunk 3]
```diff
@@ -555,7 +555,7 @@ func (pr *PullRequest) Merge(doer *User, baseGitRepo *git.Repository, mergeStyle
 		return ErrInvalidMergeStyle{pr.BaseRepo.ID, mergeStyle}
 	}
 
-	env := PushingEnvironment(doer, pr.BaseRepo)
+	env := FullPushingEnvironment(doer, doer, pr.BaseRepo, pr.ID)
```
[the remark this answers is not in the document]
Comment (zeripath) [reply models/pull.go]: Done

**Comment by zeripath** [comment]
Conflicts resolved.

**Comment by zeripath** [comment]
So the next step once we have this PR is to enforce that we have to have an appropriate Gitea environment before allowing pushing to gitea repositories - this will prevent the numerous issues that we get whereby people abuse Gitea repositories by pushing outside of Gitea.

**Review by lunny — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by stealthybox** [comment]
glad to see this fixed in gitea
thank you @zeripath :taco: 

related gogs issue #5120

**Pull request merged**